UefiPayloadPkg/Library/FdtParserLib/FdtParserLib.c

```diff
@@ -0,0 +1,627 @@
+          if (PayloadBase == NULL) {
+            return EFI_OUT_OF_RESOURCES;
+          }
+
+          PayloadBase->Header.Revision = UNIVERSAL_PAYLOAD_BASE_REVISION;
+          PayloadBase->Header.Length   = sizeof (UNIVERSAL_PAYLOAD_BASE);
+
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "addr", &TempLen);
+
+          ASSERT (TempLen > 0);
+          if (TempLen > 0) {
+            Data64       = (UINT64 *)(PropertyPtr->Data);
+            StartAddress = Fdt64ToCpu (*Data64);
+            DEBUG ((DEBUG_INFO, "\n         Property(00000000)  entry"));
+            DEBUG ((DEBUG_INFO, "  %016lX\n", StartAddress));
+
+            PayloadBase->Entry = (EFI_PHYSICAL_ADDRESS)StartAddress;
+          }
+        }
+
+        if (AsciiStrnCmp (NodePtr->Name, "upl-params", AsciiStrLen ("upl-params")) == 0) {
+          PropertyPtr = FdtGetProperty (Fdt, SubNode, "address_width", &TempLen);
+          if (TempLen > 0) {
+            Data32 = (UINT32 *)(PropertyPtr->Data);
+            DEBUG ((DEBUG_INFO, "\n         Property(00000000)  address_width"));
+            DEBUG ((DEBUG_INFO, "  %X", Fdt32ToCpu (*Data32)));
+            SizeOfMemorySpace = (UINT8)Fdt32ToCpu (*Data32);
+ #if defined (MDE_CPU_IA32) || defined (MDE_CPU_X64)
+            BuildCpuHob (SizeOfMemorySpace, 16);
+ #else
+            BuildCpuHob (SizeOfMemorySpace, 0);
+ #endif
+          }
+        }
+      }
+    }
+    // Optional
+    else if (AsciiStrnCmp (NodePtr->Name, "pci-rb", AsciiStrLen ("pci-rb")) == 0) {
+      DEBUG ((DEBUG_INFO, "  Found pci-rb node (%08X)", Node));
+
+      HobDataSize = sizeof (UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES) + RootBridgeCount *sizeof (UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGE);
+      //
+      // Create PCI Root Bridge Info Hob.
+      //
+      if (PciRootBridgeInfo == NULL) {
+        PciRootBridgeInfo = BuildGuidHob (&gUniversalPayloadPciRootBridgeInfoGuid, HobDataSize);
+        ASSERT (PciRootBridgeInfo != NULL);
+        if (PciRootBridgeInfo == NULL) {
+          break;
+        }
+
+        ZeroMem (PciRootBridgeInfo, HobDataSize);
+        PciRootBridgeInfo->Header.Length    = sizeof (UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES);
+        PciRootBridgeInfo->Header.Revision  = UNIVERSAL_PAYLOAD_PCI_ROOT_BRIDGES_REVISION;
+        PciRootBridgeInfo->Count            = RootBridgeCount;
+        PciRootBridgeInfo->ResourceAssigned = FALSE;
+      }
+
+      for (Property = FdtFirstPropertyOffset (Fdt, Node); Property >= 0; Property = FdtNextPropertyOffset (Fdt, Property)) {
+        PropertyPtr = FdtGetPropertyByOffset (Fdt, Property, &TempLen);
+        TempStr     = FdtGetString (Fdt, Fdt32ToCpu (PropertyPtr->NameOffset), NULL);
+
+        if (AsciiStrCmp (TempStr, "ranges") == 0) {
+          DEBUG ((DEBUG_INFO, "  Found ranges Property TempLen (%08X)\n", TempLen));
+
+          PciRootBridgeInfo->RootBridge[index].AllocationAttributes = EFI_PCI_HOST_BRIDGE_COMBINE_MEM_PMEM | EFI_PCI_HOST_BRIDGE_MEM64_DECODE;

```

Comment:
This whole index logic sounds very odd. Index should be tied to number of root bridges. The way it is manipulated inside like if idex = 1, make it 0 and all. It is not clear why?

Reply:
pci-rb loop has been updated to resolve index issue